### ISSUE TITLE
Bundle NextJS static export into single index.html file

### ISSUE THREAD

**Issue by kannwism** [webdiscus/html-bundler-webpack-plugin#37]
Hi everyone,

My goal is to bundle a NextJS application into a single `index.html` file. I was very happy to stumble across this plugin, as it seems very promising for this task!

What I've done:

1. Create NextJS App
2. Run Static Export -> Generates `out/` directory with the following structure:
```
- out
  - \_next
    - lvHzHE4P5UBVueHLgV9u6J
  - static
    - chunks
      - app
        - _not-found-be6954023b538636.js
        - layout-8c4f8055601f861.js
        - page-390a08b8bc838c8.js
      - pages
        - _app-2727a7117f49dcf1.js
        - _error-91a5938854a6f402.js
      - 864-5b2a0e4b041824d8.js
      - 964-ed976b209e7a86b0.js
      - fd91056d-bdc2f178762b876.js
      - framework-8883d1e9be70c3da.js
      - main-app-a5c502c4547e2c37.js
      - main-eaac6a2efdb8f260.js
      - polyfills-c67a75d1bf699dc8.js
      - webpack-bba7619efc7529ac.js
      - css
        - 8aba845b4757a42c.css
    - lvHzHE4P5UBVueHLgV9u6J
        - _buildManifest.js
        - _ssgManifest.js
    - media
      - ...
    - 404.html
    - favicon.ico
    - index.html
    - index.txt
    - next.svg
    - vercel.svg
 ```
3. Run webpack with the following config:
```
const HtmlBundlerPlugin = require("html-bundler-webpack-plugin")

module.exports = {
  plugins: [
    new HtmlBundlerPlugin({
      entry: "./out/",

      js: {
        filename: "assets/js/[name].[contenthash:8].js",
        inline: true, // inlines JS into HTML
      },
      css: {
        filename: "assets/css/[name].[contenthash:8].css",
        inline: true, // inlines CSS into HTML
      }
    })
  ],

  mode: "production",
}
```
4. The output is `dist/` directory with `404.html` and `index.html`, but unfortunately the scripts are not inlined.

The repository with everything I mentioned can be found here: https://github.com/kannwism/next-js-single-html/

I'm new to webpack so I don't know if this is a user error / misconfiguration on my side, or if this is a bug or unsupported feature of the plugin.

In any case, there might be other people interested in achieving this, e.g. on [reddit](https://www.reddit.com/r/nextjs/comments/ynk4q6/can_i_build_a_nextjs_project_in_one_html_file/).

I know that NextJS is not meant to do this, but our team has already put a lot of effort into the NextJS app.

Any help is highly appreciated!

**Comment by webdiscus** [comment]
Should absolutely all resources be inlined: font, images, svg, css, js or only CSS and JS?

I see you have preloaded fonts, images, JS.
Inlined assets should not be preloaded, because it makes no sense.

P.S. This is interest special use case.
The HTML is already generated via NextJS and you wish use the Webpack as a post process to inline all assets. With the Bundler Plugin it's possible. I need to think, how optimise this process.

**Comment by webdiscus** [comment]
@kannwism 

it's worked out, see please my [PR](https://github.com/kannwism/next-js-single-html/pull/1)

**Comment by webdiscus** [comment]
It works in 2 steps:

1 - `npm run build` - generate the `out/` directory via next.js
2 - `npm run webpack:inline` - cleanup html from unless tags and inline used assets into html

**Comment by kannwism** [comment]
Hi @webdiscus 

Wow, that was fast! Thank you very much! Indeed the PR works for the example project 👍 

Now I tried to replicate this in a much bigger NextJS project (unfortunately I can't share the code publicly). It bundles everything into .html files as desired, but when serving it and accessing in the browser I get this error in the console:

```
TypeError: Cannot read properties of null (reading 'textContent')
```

From some googling it seems that the issue is that the scripts are executing before the HTML Document has completely loaded. Before inlining, the `defer` tag would prevent this, but it's not an option for inline scripts.



Apparently one could achieve the behavior with something like this:
```
<script>
    window.addEventListener('DOMContentLoaded', (event) => {
        // original script
    });
</script>
```
(found it here: https://stackoverflow.com/questions/41394983/how-to-defer-inline-javascript )

I will try to enclose the tags that are marked with `defer` with the snipped using the `beforePreprocessor` function in the webpack config.

Let me know if you have any other ideas :)

**Comment by webdiscus** [comment]
Hi @kannwism,

_Just for your information:_
Using the DOMContentLoaded event is usually practice to run a script after DOM completely loaded. But the DOM is not yet rendered, it means that the DOM elementen are selectable, but the position and size of elements is yet not calculated, simply to say - CSS is yet not applied.

To be sure that DOM is loaded and rendered you can use the snipped:
```js
window.addEventListener('DOMContentLoaded', (event) => {
  // DOM is loaded, but yet not rendered
  // here you can change the DOM: add/remove/delete DOM elements
  
  requestAnimationFrame(() => {
    // DOM is loaded and rendered,
    // here you can get real size and position of DOM elements
  })
});
```

**Comment by webdiscus** [comment]
> do you have any idea how one could enclose the scripts that have a defer tag with the snippet?

hmm, I can add this functionality to wrap the source code in _onReady_ snipped, but:

- this is a _black magic_, bad practice
- the source code must be analysed whether the code is already wrapped with _onReady_, but it's impossible, because the onReady code can be as a function with any name...
- you should serve your JS code yourself

Add `onReady` JS code сниппет to each of your scripts yourself.
This should happen on the developer's side, and not on the plugin's side.

**Comment by webdiscus** [comment]
@kannwism 

> I will try to enclose the tags that are marked with defer with the snipped using the `beforePreprocessor` function in the webpack config.

The `beforePreprocessor` function is called before all processes, the html contains you original code where assets are not yet inlined.

I'm just working on the next version where will be added a new callback and hook - `afterCompile`. In this hook will be avaliable fully compiled HTML with inlined assets.  The HTML can be modified in this hook, right before it will be saved on disk.

**Comment by webdiscus** [comment]
@kannwism 

Thank You for donation!
Your name is now on the sponsor list in the readme.

Grüß aus Köln

**Comment by kannwism** [comment]
Hey @webdiscus 

One of the issues is that there are scripts with the `defer` option and some without. And once the scripts are inlined the information is lost, whether the script needs deferring or not.

Secondly, I tried to recreate the problem in the [basic NextJS app](https://github.com/kannwism/next-js-single-html) by adding a little functionality. E.g. a text field and a button to update text on the page. Now I don't get any errors, but the functionality doesn't work after the webpack step...

So basically `npm run dev` works as expected, but `npm run build:inline` breaks the functionality, without throwing errors.

> Thank You for donation!
Your name is now on the sponsor list in the readme.

You're welcome :)

Grüsse aus Zürich

**Comment by webdiscus** [comment]
> Secondly, I tried to recreate the problem in the [basic NextJS app](https://github.com/kannwism/next-js-single-html) by adding a little functionality. E.g. a text field and a button to update text on the page. Now I don't get any errors, but the functionality doesn't work after the webpack step...

I'll check, it's tonight

**Comment by webdiscus** [comment]
> One of the issues is that there are scripts with the defer option and some without. And once the scripts are inlined the information is lost, whether the script needs deferring or not.

_This is not a bug, this is the feature_ ;-)

The Inlined script code is enclosed in the `<script>... JS code ...</script>` tag.
The inline `<script>` tag cannot have any attributes.

[MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script?retiredLocale=de#defer):

> **Warning**
>
> The `defer` attribute must not be used if the src attribute is absent (i.e. for inline scripts), in this case it would have no effect.

But, I can implement the feature to keep original attributes with the  inline script tag.
The feature can be enable/disable via the [`js.inline`](https://github.com/webdiscus/html-bundler-webpack-plugin#js) option, e.g., (suggestion only):

```js
js: {
  inline: {
    keepAttributes: true | false, // default will be false
  },
},
```

The source HTML:

```html
<script src="javascript.js" defer="defer"></script>
```


Then the processed HTML will contains the inline script tag with original attributes, except `src`:
```html
<script defer="defer">
  console.log('Hello World!');
</script>
```

Of cause in the `afterCompile` (yet not implemented) you can modify/remove the script tags in HTML.

It's would be ok?

**Comment by kannwism** [comment]
> js: {
  inline: {
    keepAttributes: true | false, // default will be false
  },
},

Yeah something like this would be helpful, then I could probably even write a little custom plugin to remove the defer and wrap the script in an eventListener :)

Only thing is I'm not sure if that actually solves my NextJS issue, since I don't quite understand what the problem is that I got in the sample app 😆. So maybe I should figure that out before you go through the effort to make a custom option.

**Comment by webdiscus** [comment]
Hallo @kannwism,

I have a beta version `2.16.0-beta.1` with paar new features.

### Added the callback `js.inline.keepAttributes` option:

For example, there is script with the attribute:
```html
<script src="javascript.js" defer></script>
```

Add the `keepAttributes` callback option:
```js
   new HtmlBundlerPlugin({
      entry: {
        index: './src/index.html',
      },
      js: {
        inline: {
          // return true to keep the attribute
          // return  false of undefined (no return) to remove an attribute
          // defaults (when the keepAttributes is not defined) all attributes will be removed
          keepAttributes: ({ attributes, attribute, value }) => {
            if (attribute === 'defer') return true;
          },
        },
      },
    }),
```

Generated HTML contains allowed attibute:
```js
<script defer>console.log("Hello World!");</script>
```
Do you have any suggestions?

###  Added `beforeEmit` callback:
```js
new HtmlBundlerPlugin({
      entry: {
        index: './src/index.html',
      },
      /** {string} content The generated html before save to file. You can return modified html. */
      beforeEmit: (content) => {
        return content.replace('REPLACE ME', 'The new replaced content.');
      },
    }),
```

## HOOKs

You can create own plugin for the bundler plugin using follow hooks:

- `beforePreprocessor` called before the source template will be processed
- `beforeEmit` called after processing, before save to file, generated html contains all output asset filenames and inlined JS/CSS. You can return modified html.
- `afterEmit` called after html is saved, you can't change the html, but you get the dependencies tree (hierarchy) of all processed assets, the root asset is the generated html file


## P.S.

All features are `experimental` and can be renamed/changed in the `beta` version.
The experimental features are undocumented, because it's yet `beta`.
The specification you can see in the `types.d.ts`. 

The beta version is just to try it out and give a feedback/suggestions.

**Comment by webdiscus** [comment]
## callbacks vs hooks

The callbacks and hooks called at the same place in code and function is the same.

### When using `callbacks`
When you have small code just for your project only. You can use it for debug or small fixes.

### When using `hooks`
When your code is big and/or your code can be useful for other projects, e.g. inject in head auto generated favicon tags for all media.

**Comment by webdiscus** [comment]
@kannwism 

> So basically npm run dev works as expected, but npm run build:inline breaks the functionality, without throwing errors.

I know why... 
The next.js has very complex default config and will split JS files into many dynamic loaded chunks, which is not inlinable, because already inlined before body tag:
```js
...
<script>
self.__next_f.push([
1,
  '4:I{"id":6054,"
chunks":[
"272:static/chunks/webpack-5644af036015bd4c.js", // <= dynamic chunks
"971:static/chunks/fd9d1056-d9a18e37b76d597b.js", // <= dynamic chunks
"864:static/chunks/864-b9e8988d69d5d1ed.js" // <= dynamic chunks
],
]);
</script>
</body>
```

This dynamic chunks is impossible to inline into HTML, I'm sorry. It's very complex.

## Workaround
You can change the next.js default config to disable `splitChunks` (it's make no sense, because js will be inlined).
